Move @userplot to RecipesBase

ChrisRackauckas opened the pull request:
It makes sense. It's used in libraries which use recipes but don't want to depend directly on Plots.jl. For example: PlotRecipes.jl.

mkborregaard commented:
Yes, I have thought about this too but doesn't it cause problems that the function depends on `plot` since what it does is just define a new method for that? If it were to be moved, `shorthands` should be moved too, IMO.

ChrisRackauckas commented:
I'm not entirely sure, but a discussion for how to make this possible surely should happen. I thought it generated the plot commands on demand so it could be okay?

mkborregaard commented:
Would you be willing to test how `@userplot` behaves if you insert it into a module that depends on RecipesBase but not Plots?

ChrisRackauckas commented:
Not quite right now.

daschw commented:
I tested `@userplot` behavior with the following module:
```julia
__precompile__()

module TestPlots

using RecipesBase

@userplot RandomPlot

@recipe function f(rp::RandomPlot)
    randn(10)
end

end
```
Originally everything worked fine. After applying the changes, @ChrisRackauckas proposed, (`export @userplot` would have to be moved from Plots to RecipesBase as well, btw), I got the following error:
```julia
julia> using TestPlots, Plots

julia> randomplot()
ERROR: UndefVarError: plot not defined
 in #randomplot#1(::Array{Any,1}, ::Function) at /home/dani/.julia/v0.5/RecipesBase/src/RecipesBase.jl:319
 in randomplot() at /home/dani/.julia/v0.5/RecipesBase/src/RecipesBase.jl:319
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
```
So, I guess @mkborregaard was right with his doubts concerning the `plot` dependency.

ChrisRackauckas commented:
Is there a way to modify `@userplot` to remove its dependence on `plot`?

Evizero commented:
why not put the definition `function plot end` into `RecipesBase.jl` and have `Plots.jl` import it

ChrisRackauckas commented:
yeah, that would do it.

mkborregaard commented:
Won't putting `plot` in the workspace with `using RecipesBase` annoy package owners that wish to define a recipe but also cater to plotting with other plotting packages?

ChrisRackauckas commented:
>Won't putting plot in the workspace with using RecipesBase annoy package owners that wish to define a recipe but also cater to plotting with other plotting packages?

Only if you export `plot`. Instead, it can define it without exporting it, and Plots.jl can `imoprt RecipesBase: plot`.

ChrisRackauckas commented:
I setup the `plot` command as discussed.

mkborregaard commented:
have you tested it?

ChrisRackauckas commented:
Not yet. I'll give it a go soon.

Evizero commented:
sorry tom, I misclicked

mkborregaard commented:
As an aside, perhaps, have you seen any use of `@recipe function f(::Type{Val{:myplotrecipename}}, plt::Plot; ...) `?  I searched the entire Plots ecosystem for the signature, but haven't found it.

ChrisRackauckas commented:
In StatPlots.jl. I'm fixing that right now. One sec.

ChrisRackauckas commented:
I moved the 3 top level abstract types to RecipesBase, and changed the apply recipe error checking to the abstract type, so now

```julia
@recipe function f(::Type{Val{:myplotrecipename}}, plt::AbstractPlot; ...)
```

works only using RecipesBase.jl. 

StatPlots.jl still needs Plots.jl as a dependency because it uses some function `cycle`. This shadows the Base function `cycle`, and I'm not sure what it actually does. But it seems important because it's used quite a bit. It can just be renamed and declared in RecipesBase.jl and then all of the recipes libraries will no longer depend on Plots.jl

Oh and this fixes the Media precompilation problems. @sbromberger may want to give this a try.

ChrisRackauckas commented:
The only thing that would be needed for `cycle` to work as is, but without shadowing Base, would be to `import Base: cycle` in Plots. All of the dispatches are on Plots-specific types, so those are okay, except for one:

```julia
cycle(v, idx::Int)
```

Would this be allowed or is this a form of piracy? @tkelman

tkelman commented:
base method, base types...

ChrisRackauckas commented:
That's what I thought. So change the name to `_cycle`?

mkborregaard commented:
Does it make a difference that the method isn't exported?

ChrisRackauckas commented:
>Does it make a difference that the method isn't exported?

It would still change the behavior of `Base.cycle`, even if `Plots` does not export it.

mkborregaard commented:
It's not really changing the behaviour, as there is no two-argument method in Base, though?

mkborregaard commented:
Anyway, if that is the case, I think the rename to _cycle is still prudent. We shouldn't shadow a Base function.

mkborregaard commented:
@pkofod @daschw @Evizero @jheinen @SimonDanisch This is a big proposed reorganisation, I think it'd be great to exchange opinions before going ahead with it. @tbreloff has indicated that he is happy for us to go forward if it doesn't break anything.
My main concern is that it should be tested and solid before merging, as RecipesBase should preferably change rarely (as far as I can see, though, the only change that will break anything downstream is renaming `cycle`).

Evizero commented:
I personally like the change. I do recall however that the idea of moving `plot` to this package came up before in a conversation I was having with Tom. I forgot where it was but he had some reservations back then but was open to the general idea.

> My main concern is that it should be tested and solid before merging

Hard to disagree with that

pkofod commented:
Seem sensible. I'm all for making it easier to adopt `Plots` without putting `Plots`in `REQUIRE`.

daschw commented:
I also like the change.
@Evizero I think you are talking about the discussion with Tom here https://github.com/JuliaPlots/RecipesBase.jl/issues/12, right?

Evizero commented:
@daschw Yes that was it. Good detective work!

ChrisRackauckas commented:
Alright, current status. This branch now has the `_cycle` and `@shorthands` which are used quite commonly. PlotRecipes has no Plots.jl dependency, and all of the recipes can be used from RecipesBase. StatPlots.jl still has a Plots.jl dependency, because there's a few more functions that we would need to find out what to do with. I think they may be better suited for PlotUtils?

"This change" is this branch plus:

https://github.com/JuliaPlots/StatPlots.jl/pull/50
https://github.com/JuliaPlots/PlotRecipes.jl/pull/21
https://github.com/JuliaPlots/Plots.jl/pull/786

The remaining Plots functions in StatPlots are:

```
Plots.discrete_value!
Plots.@deps
Plots.fg_color
Plots.extractGroupArgs
Plots.isvertical
Plots.istuple
```

I have Plots.jl still as a dep on StatPlots.jl so this can be tested, but would like to see if we can go all the way.

ChrisRackauckas commented:
`istuple` is unnecessary: just switched to `typeof(x) <: Tuple` in that one spot. So now we're down to 5 functions left to handle.

ChrisRackauckas commented:
References:

`discrete_values!`: https://github.com/JuliaPlots/Plots.jl/blob/731cac42620a06e2197fea1132a3d4990a8d8a8f/src/axes.jl#L423

`@deps`: https://github.com/JuliaPlots/Plots.jl/blob/a3b4b9c8b28262e11cd911ebfb20993ce9406dd5/src/recipes.jl#L79

`fg_color`: https://github.com/JuliaPlots/Plots.jl/blob/master/src/args.jl#L1007

`extractGroupArgs`: https://github.com/JuliaPlots/Plots.jl/blob/master/src/args.jl#L760

`isvertical`: https://github.com/JuliaPlots/Plots.jl/blob/master/src/utils.jl#L395

Note these didn't show up in the original catch for StatPlots because it was only `import Plots: cycle`, these other ones were only as `Plots. ....` 👍

tbreloff commented:
I have to ask... why are you pushing this so hard? Not only is it against
the spirit of RecipesBase to throw all this stuff in there, then one can no
longer call "using StatPlots" (or PlotRecipes) and expect it to work. Lots
of existing code will break without the Plots dep I think.

Anyways I just wanted to add my $0.02, but feel free to push Plots in
whatever way is best for the community. I love the effort, and sorry I
don't have the time to help.

On Mon, May 1, 2017 at 8:18 PM Christopher Rackauckas <
notifications@github.com> wrote:

> istuple is unnecessary: just switched to typeof(x) <: Tuple in that one
> spot. So now we're down to 5 functions left to handle.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaPlots/RecipesBase.jl/pull/16#issuecomment-298466250>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA492jouTGAOHMcf7_t613W4PzAbwkH-ks5r1nZFgaJpZM4NMKEV>
> .
>

ChrisRackauckas commented:
>I have to ask... why are you pushing this so hard? 

I would like the ability to make recipes without requiring Plots.jl. This is the promise of RecipesBase. But in actual use, it's quite limited. Not all recipes can, in the current architecture, be implemented without a Plots.jl dependency.

Not wanting the Plots.jl dependency is two-fold. For one, Plots.jl doesn't always work or precompile on every setup. This is because it's a large library. For example, right now it's not working on v0.6. One dependency not working will take down a whole Julia package, and so for stability reasons avoiding a large library which links to many dependencies which can cause problems is an issue. But secondly, there is a precompilation problem. It's now limited to only the case where a precompiled library depends on Plots, but it's pretty consistent now:

https://github.com/JuliaPlots/PlotRecipes.jl/issues/20#issuecomment-294594986

@dlfivefifty ran into this when Plots.jl turned back on precompilation, and he hit the Media bug. The only answer is to remove Plots.jl. His response was to remove all userplot recipes since those require Plots.jl. Instead, after this change, he could use userplot recipes again just through RecipesBase.

>Not only is it against the spirit of RecipesBase to throw all this stuff in there

RecipesBase is actually not growing much at all with the current changes. The reason why I haven't done anything with those 5 functions is because I am not sure what to do with them to cleanly remove their usage from StatPlots or move them somewhere else (likely PlotUtils.jl?). This is still a tiny library: I moved essentially 2 macros, one of which was small. There will still be people who complain about a dependency because they are too silly to actually check the library instead of counting dependency numbers, but those people probably wouldn't accept the dependency even if the library was 20 LoC (see UnicodePlots.jl...) so I don't really see a practical difference here.

>Lots of existing code will break without the Plots dep I think.

No libraries use these.

````
julia> Pkg.dependents("PlotRecipes")
0-element Array{AbstractString,1}

julia> Pkg.dependents("StatPlots")
1-element Array{AbstractString,1}:
 "PlotRecipes"
```

I think the Media bug prevented the proliferation of these libraries quite severely, so it's only seen in end-user code.

>then one can no longer call "using StatPlots" (or PlotRecipes) and expect it to work.

Was that ever supposed to be the case for recipes?

ChrisRackauckas commented:
I think there's two separate things here. The first thing is: should all recipes be able to be defined using only RecipesBase? That answer I think is an easy yes, and the simplest version of these fixes (move macros and abstract types here) does that. I believe that one is much less controversial, so that should be answered first.

That actually allows PlotRecipes.jl to be Plots.jl free, but not StatPlots.jl. So the next question I was investigating was, how hard would it be for StatPlots.jl to be Plots.jl free? Two reasons. One, I was wondering how much of the Plots.jl internals it was really using: how possible is it to do things like StatPlots.jl in a library that can't depend on Plots.jl? But secondly, it would make StatPlots.jl usable by other libraries.

However, given what I have seen now by trying to do it, finishing up the change to make StatPlots.jl Plots-free might be both more difficult that I imagined at first. There's 4 functions and 1 macro that seems to require some of Plots.jl's internals. If this can't be done, I am fine it just giving up this attempt and reverting back to just "RecipesBase for all recipe types" as a fine result on its own.

mkborregaard commented:
I've spent some time thinking about this, and here's what I think.
The integrity of RecipesBase is of primary importance. I see this as the key advance of Plots. There are lots of good plotting packages, and Plots has many advantages, e.g. that all plots are generated with the same suite of keywords making it easy to reason about comments, the succinct syntax made possible by the series concept, etc. But the key is to be able to write code that will plot with a number of different plotting packages, without depending on any of them, and without disrupting the user's ability to choose a different way of plotting. This makes it possible even for basal library code to define plotting methods without issues.

You all know that, but I'm just stressing it because I think it's so important. This means that 1) changes in RecipesBase can't break anything downstream, and 2) once we add something to RecipesBase we cannot take it out later. I think it is key that we don't tempt/force users to specify version numbers for RecipesBase dependencies. Right now it's essentially a free dependency. It puts two macros in the workspace that doesn't conflict with most other functionality or other plotting packages, it has no dependencies itself, AND it can never cause version conflicts and dependency hell downstream because it is fairly static.

I worry about putting `plot`in the workspace in particular - even if it isn't exported, it is still made available when calling the userplot macro, and the possibility for this to cause issues for downstream packages using other plotting packages should deter most basal package owners from depending on RecipesBase.

However, I do see the idea of being able to make more advanced recipes using the entire recipes arsenal as something useful. Here is one suggestion that could perhaps solve all of these issues: What about a new package ("RecipesTools", "AdvancedRecipesBase", "RecipesBasePlus") that reexports RecipesBase and defines the advanced macros (`@shorthands`, `@userplot` and the `plot` function, probably `@layout`, but maybe not `_cycle`). Plots itself would then depend on that package instead of RecipesBase. This would solve the issue of versions with RecipesBase while opening the possibility of more advanced recipes at the cost of possibly a larger commitment to the Plots ecosystem for package owners.

With regards to StatPlots and PlotRecipes I don't think it is worth ridding them of the Plots dependency. They currently serve two uses 1) defining plots for things that require lots of heavy dependencies - they can be used by users who want that functionality at the cost of making them less stable; and 2) to define and document specialized plot functions with different names (e.g. `graphplot`) that use non-standard keywords that don't fit within the classic Plots 'attributes'. Especially the first use means that I think it is not worth the cost to try to rid the Plots dependency. I think it could be argued to split off the second functionality in a package that only depends on RecipesBase and nothing else as I think @ChrisRackauckas has argued for, but that is a slightly different discussion.

I think  there is a bit of misunderstanding about what these packages do - I believe they are meant as higher level inclusive packages (like Learn.jl in the JuliaML ecosystem), so they are not meant just as a collection of recipes (maybe the PlotsRecipes name is a bit unfortunate). Users are expected to just `using StatPlots` and forget about Plots - not `using Plots, StatPlots` etc. So I suggest not changing those packages.

Evizero commented:
`RecipesTools` 👍

tkelman commented:
> I think it is key that we don't tempt/force users to specify version numbers for RecipesBase dependencies

What? Why not? If you add features here that haven't existed in every version that's ever been released, and people want to use them, then they should be marking the dependency with a minimum version accordingly.

mkborregaard commented:
Yes absolutely. It'd be great if you could disprove my concerns @tkelman . Here is what I mean: All packages should be able to depend on RecipesBase without penalty. Many package creators of basal packages are wary of dependencies, because of the vulnerability to bugs if the dependency breaks (we have that issue quite intensively with PlotRecipes right now), and because of the potential for version conflicts, if several basal packages use a different version of each package.

The idea of RecipesBase is to have something that is absolutely safe to add as a dependency that will still allow you to define plotting functionality. I think RecipesBase is the best solution I've seen to deal with julia's design with custom types (that can dispatch plotting functionality) together with a very distributed functionality for plotting.

So yes, if we add features here that haven't existed in every version that's ever been released the user should use version numbers. And that exactly may create a potential for version conflict. Especially if some of that functionality is removed later, adding the necessity of upper version bounds.

ChrisRackauckas commented:
>So yes, if we add features here that haven't existed in every version that's ever been released the user should use version numbers. And that exactly may create a potential for version conflict. Especially if some of that functionality is removed later, adding the necessity of upper version bounds.

>You all know that, but I'm just stressing it because I think it's so important. This means that 1) changes in RecipesBase can't break anything downstream, and 2) once we add something to RecipesBase we cannot take it out later. I think it is key that we don't tempt/force users to specify version numbers for RecipesBase dependencies. Right now it's essentially a free dependency. It puts two macros in the workspace that doesn't conflict with most other functionality or other plotting packages, it has no dependencies itself, AND it can never cause version conflicts and dependency hell downstream because it is fairly static.

The two plot recipes that I am proposing making available here requires the movement of 3 abstract types, one non-breaking syntax change (dispatch on the abstract type in the recipe if you are only depending on RecipesBase), and moves one macro here. This makes two out of the four recipes which have been available in the docs for over a year available from RecipesBase. Do you think that this reduced PR is on a stable enough part of the infrastructure to be in here?

I like the idea of a RecipesTools to start "finishing" the Plots-less recipes, i.e. the rest of this PR can go there.

mkborregaard commented:
Thanks for moving so fast on this - I'd feel comfortable with this IF I felt that we had established beyond doubt that calling `@userplot` in a package depending on RecipesBase wouldn't get name clashes on `plot` if they are also replying on e.g. Gadfly or PyPlot proper.

tkelman commented:
> Especially if some of that functionality is removed later, adding the necessity of upper version bounds.

One thing I recommend doing for this kind of deprecation cycle is timing the first version that has deprecated functionality removed to be the same as the first version of the package that drops support for some old Julia version. At least pre-1.0, old Julia vs new Julia is a pretty good way of separating packages that want to keep using older functionality from things that are moving more in step with their upstream dependencies.

mkborregaard commented:
That's a good point.

mkborregaard commented:
I guess that means now is the time to make the decision.

dlfivefifty commented:
I think creating new packages to avoid breaking changes is not a sustainable model for version control. So unless there's a good argument for RecipesTools existing on its own, a normal deprecation update would be preferable.

While the dependencies breaking the whole ecosystem is incredibly annoying, keep in mind we are talking about software (both Julia itself and packages) with version starting with 0.*, so users at this point should probably be considered beta testers.

mkborregaard commented:
Those are good points. To explain a bit further 

> dependencies breaking the whole ecosystem is incredibly annoying

I agree that breaking the ecosystem is in principle OK, I just want if possible to preserve RecipesBase as a 'safe' dependency. Although I think aligning with julia versions as suggested could fix this.

> unless there's a good argument for RecipesTools existing on its own

Yes, that is the key question. There is only an argument for it if `@userplot` does in fact cause trouble by introducing `plot` to the namespace. With the big selection of plotting packages in Julia, adding RecipesBase as a dependency should never commit the user to use Plots. Not quite sure how to test this satisfactorily.

ChrisRackauckas commented:
>There is only an argument for it if @userplot does in fact cause trouble by introducing plot to the namespace. 

Why do you think it does that? This doesn't export `plot`, it just defines it here for Plots.jl to include it. If it's something to do with the internals of the macro (from I can tell, it doesn't define a plot command either, just functions which call plot), we can change those to `RecipesBase.plot`. The `plot` command needs to be defined here, but not `export`ed to the user at all.

mkborregaard commented:
I see what you mean. I have tested out your patch and also various other plotting packages and there doesn't seem to be an issue. And the general consensus seems to support merging!

So let's do it. Let's tag a release, then merge and up the version requirement to 0.6 and do a 0.6 release. @ChrisRackauckas is this PR in your opinion completely finished? What about `layout`, while we're at it?

ChrisRackauckas commented:
>What about layout, while we're at it?

```julia
function create_grid(s::Symbol)
    :(EmptyLayout(label = $(QuoteNode(s)), blank = $(s == :_)))
end

macro layout(mat::Expr)
    create_grid(mat)
end
```

and then

```julia
# EmptyLayout

# contains blank space
type EmptyLayout <: AbstractLayout
    parent::AbstractLayout
    bbox::BoundingBox
    attr::KW  # store label, width, and height for initialization
    # label  # this is the label that the subplot will take (since we create a layout before initialization)
end
EmptyLayout(parent = RootLayout(); kw...) = EmptyLayout(parent, defaultbox, KW(kw))

Base.size(layout::EmptyLayout) = (0,0)
Base.length(layout::EmptyLayout) = 0
Base.getindex(layout::EmptyLayout, r::Int, c::Int) = nothing

_update_min_padding!(layout::EmptyLayout) = nothing
```

.... seems like it keeps going. I think I'd have to pull in the entire layout system to get that.

But any layout can be specified without the macro, right? So I think we're fine, and in the future the `@layout` macro could be re-written to be just a syntax translation that doesn't rely so heavily on the internal types, and then it could go here. But as it stands now, it works by directly calling one of the core concrete type and uses that to do the syntax translation, and untangling all of that is a whole different ordeal.

ChrisRackauckas commented:
>@ChrisRackauckas is this PR in your opinion completely finished? 

Let me get all of the PRs completely finished, including the docs PR for the abstract type in the recipe. I'll post here when that's done.

mkborregaard commented:
Perfect, thanks!

ChrisRackauckas commented:
Alright, I pushed the backtracked versions. This PR no longer includes `_cycle` because that was to try and get Plots out of StatPlots which is a much larger ordeal. 

https://github.com/JuliaPlots/Plots.jl/pull/786

Is part of the move, and also does the rename to `_cycle` because of the previously mentioned confusion (but no longer is that defined in RecipesBase, so that's kind of a separate issue now... but can it ride along?)

https://github.com/JuliaPlots/StatPlots.jl/pull/50

is now just some QoL changes + `_cycle` rename.

PlotRecipes PR closed since it was no longer necessary.

https://github.com/JuliaPlots/PlotDocs.jl/pull/24

does the docs change so that way the recipes can be used with only RecipesBase.

So the scope is much much smaller now.

mkborregaard commented:
Yeah I don't have VisualRegressionTests working on my system, which is a bit of a nightmare because those are the important ones.

mkborregaard commented on src/RecipesBase.jl:
```diff
@@ -3,10 +3,27 @@ __precompile__()
 
 module RecipesBase
 
+using Compat
```
We'll make this 0.6 only, so no need to use compat.

mkborregaard commented:
@tkelman I'd really appreciate your opinion here. The intention of this PR is to increase the capabilities of RecipesBase somewhat, by defining a few more names in it. Specifically it defines the unexported functions `plot` and `animate` with no implementation (i.e. `function plot end`) and two macros that when run in a package will parse to the code `fooplot(x...) = RecipesBase.plot(fooplot, x...)`. My concern is the potential for name clashes if users use a different plotting package, such as PyPlot or Gadfly.

Here's the problem: the RecipesBase package is meant to be a no-conflicts dependency. In principle, libraries like StatsBase should be able to import it, without ever causing issues downstream, for dependent packages or users who have no wish to use Plots. Right now, RecipesBase only exports a macro that defines a method for the `apply_recipe` function (defined in Plots). This will never cause conflicts.

In your opinion, is it unproblematic to define plot and animate (and the macros) as detailed above, or will it increase the conflict potential of the package? Thank you!

tkelman commented:
That is a very good question. Would be good if we had a method-merging feature to make this less problematic. If RecipesBase is a small, well-maintained, low-cost dependency, then I guess you could attempt to make it the agreed-upon home for those generic functions if other packages would be willing to go along with that and import it? The rest of the package with plot recipes is kind of Plots.jl-specific though, so it's maybe not the ideal place for that. There is also Graphics.jl that could be used for this purpose?

ChrisRackauckas commented:
There will never not be an issue with the `plot` function because all plotting libraries add a lot of the same dispatches like `plot(::AbstractMatrix)`. 

My proposed move does not `export plot`, so it shouldn't create any silent clashing. I user would have to explicitly `import RecipesBase: plot`.

mkborregaard commented:
Thanks for the prompt response!
The best solution for us would be to make RecipesBase a base dependency for all plotting packages - It's only 285 lines of code and has zero dependencies, and is meant to practically never have to be changed. I do doubt however that we could gather consensus for that.

Given that Graphics.jl imports NaNMath and Colors it's much less light, so I think it would be preferable not to have to depend on it (although these packages are pretty ubiquitous).

There are two design options that avoid this, but will make the package more atomized: 1. to define a base plotting package that only contains e.g. `function plot end` and `function animate end` (or have that in Base); or 2. to leave RecipesBase as is, but put @ChrisRackauckas ' PR in a new package importing RecipesBase for those who wants the extra functionality.

mkborregaard commented:
Thanks for the comments @ChrisRackauckas , as you know I'm positive towards this change but I really want to make sure we have an overview of the consequences.

mkborregaard commented:
But, @tkelman if your response is "not quite sure" that motivates me to take the path of caution and find a different solution.

ChrisRackauckas commented:
You would only need method merging if this exported `plot`, which it doesn't. You'll only get clashes if the user `using Plots, PyPlot`, and for the same reasons as before. And I think that's unavoidable. But `using RecipesBase, PyPlot` doesn't have an issue.

mkborregaard commented:
I'm happy to merge (after a rebase) if @tkelman can confirm that assertion.

tkelman commented:
Yeah, not exporting it should be safe.

ChrisRackauckas commented:
I rebased the PRs.

mkborregaard commented:
Thanks, @tkelman !
And thanks, @ChrisRackauckas , I'll merge them starting here -> Plots -> StatPlots as soon as tests pass, then release the first 0.6-only versions of all three based on that.

mkborregaard commented:
Can you update the travis testing version to 0.6?

dlfivefifty commented:
👍

tkelman commented on appveyor.yml:
```diff
@@ -1,7 +1,7 @@
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.4/julia-0.4-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.4/julia-0.4-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.4/julia-0.6-latest-win32.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.4/julia-0.6-latest-win64.exe"
```
/0.6/

tkelman commented on REQUIRE:
```diff
@@ -1 +1 @@
-julia 0.5-
+julia 0.6
```
use 0.6.0-pre or one of the rc's as a lower bound if you want it to be installable now

tkelman commented:
don't at-mention usernames in commit messages, it pings that person any time anyone pushes a rebase that includes that commit
